### PR TITLE
Master deliveryvolution rim

### PR DESCRIPTION
**Pull request by rim-odoo**
(don't merge without the enterprise branch)

### CODE REVIEW

File: addons/delivery/models/delivery_grid.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        total = order.currency_id.with_context(date=order.date_order).compute(total, order.company_id.currency_id)
+
+        return self._get_price_from_picking(total, weight, volume, quantity)
+
+    def _get_price_from_picking(self, total, weight, volume, quantity):
+        price = 0.0
+        criteria_found = False
+        price_dict = {'price': total, 'volume': volume, 'weight': weight, 'wv': volume * weight, 'quantity': quantity}
+        for line in self.price_rule_ids:
+            test = safe_eval(line.variable + line.operator + str(line.max_value), price_dict)
```
Comment (xmo-odoo) [inline addons/delivery/models/delivery_grid.py]: ???

Comment (rim-odoo) [inline addons/delivery/models/delivery_grid.py]: Nothing new under the sun: https://github.com/odoo/odoo/blob/8.0/addons/delivery/delivery.py#L237

**Comment by pedrobaeza** [comment]
Hi, @rim-odoo, we coincide again! I'm seeing that you are improving the delivery module. I was hit in the past by the removal of the table `delivery.grid` to have different prices depending on destination, as on backend (not for e-commerce, which is great), you only want to select one carrier,  and then let the system to compute the rate according customer address. My workaround was to develop the module at https://github.com/OCA/delivery-carrier/pull/113 taking advantage of a pseudo hierarchical relation on the existing table.

First question is: do you have planned to change the destination management? Maybe the principle of this module can be applied on core?

Second question is if you want me to review this PR.

Thanks.

**Comment by rim-odoo** [comment]
@pedrobaeza For this refactoring, it is rather technical cleaning than adding/changing features.

If I understand you well, you want to have an auto-selection of the delivery method according to address criteria? (in backend, in the same way it is done in the ecommerce). 

This kind of feature is not planned for now (I'm not deciding anyway) but if you want to develop it for Odoo 11 as an OCA module, you could add a button (or override the button `Get price`) in the sales order that would use `_match_address()` in order to select the delivery method. I think that would be rather easy :)

**Comment by pedrobaeza** [comment]
@rim-odoo Thanks for the explanations and the advice. Unfortunately, that's not enough, because the problem is that you can have several delivery methods for the same address, and there's other criteria for deciding which one to use (for example, the price/delivery time ratio), so the only solution would be a dynamic domain that restricts the carriers instead. Do you consider something like this?

**Comment by rim-odoo** [comment]
@pedrobaeza Not in standard nope :( but that could be a good community module?

**Comment by pedrobaeza** [comment]
Yeah, I have to think then which approach to take:

* Have the hierarchy that I have already implemented with dummy parent carrier so it's easier to select. This has no drawbacks for now in my experience and it's inheritable:
  * UPS:
    * UPS - Spain
    * UPS - France
    * ...
  * DHL
    * DHL - Spain
    * ...
  Your selection in the SO will be:
  * UPS
  * DHL
* Have that dynamic domain, which only shows the "correct" carriers. The problem is the carriers management is going to be a bit uglier (you see a lot instead of only 2 for the example). This also has the problem that if you change the delivery address, the carrier will be invalid, and if we reset it, then the user may forget to select again the appropriated. With the other approach, the same parent carrier remains selected and if there's a child carrier for the new address, the rate will be computed. If not, an error will be thrown.
  * UPS - Spain
  * DHL - Spain

Which one do you see as the most suitable?

**Pull request merged**